staging/src/k8s.io/kube-scheduler/config/v1beta1/types_pluginargs.go

```diff
@@ -77,14 +77,20 @@ type PodTopologySpreadArgs struct {
 	metav1.TypeMeta `json:",inline"`
 
 	// DefaultConstraints defines topology spread constraints to be applied to
-	// pods that don't define any in `pod.spec.topologySpreadConstraints`.
-	// `topologySpreadConstraint.labelSelectors` must be empty, as they are
-	// deduced the pods' membership to Services, Replication Controllers, Replica
-	// Sets or Stateful Sets.
-	// Empty by default.
+	// Pods that don't define any in `pod.spec.topologySpreadConstraints`.
+	// `.defaultConstraints[*].labelSelectors` must be empty, as they are
+	// deduced from the Pod's membership to Services, ReplicationControllers,
+	// ReplicaSets or StatefulSets.
+	// If empty, the default constraints prefer to spread Pods across Nodes and Zones
 	// +optional
 	// +listType=atomic
-	DefaultConstraints []corev1.TopologySpreadConstraint `json:"defaultConstraints"`
+	DefaultConstraints []corev1.TopologySpreadConstraint `json:"defaultConstraints,omitempty"`
+
+	// DisableDefaultConstraints allows to disable .defaultConstraints.
+	// Defaults to false. When set to true, DefaultConstraints must be empty
+	// or nil.
+	// +optional
+	DisableDefaultConstraints bool `json:"disableDefaultConstraints,omitempty"`
```

Comment:
Can you flip the polarity, e.g. `UseDefaultConstraints` defaulted true?

Reply:
After reading your implementation, I think if we were starting from scratch, you'd actually want a union/one-of?

E.g. the discriminator (this field) is "DefaultingType" and can be one of "SystemDefault", "List", or "None". The DefaultConstraints list may only be set if this field is "List".

Since there's existing files we want to not break (right?), I think we can default this discriminator to "List" if the list has items in it, or to "SystemDefault" otherwise. I hate to do this but I think it's OK since it's for backwards compatibility, and this is just a config file.

Reply:
I like DefaultingType. Is this just a regular enum that you called union because of its interpretation?

> Since there's existing files we want to not break (right?), I think we can default this discriminator to "List" if the list has items in it, or to "SystemDefault" otherwise.

It's alpha, so I assume it's fine to break (with the appropriate release note). However, is it normal to have to be this verbose?

Reply:
Yeah, we're eventually going to get unions as first-class citizens of the API.

It's a little verbose but it makes it very clear what you want the system to do.

Reply:
Actually, isn't "None" unnecessary? "List" + empty slice should be enough.